src/nbl/video/utilities/IUtilities.cpp

```diff
@@ -669,15 +163,662 @@ void IUtilities::updateImageViaStagingBuffer(
+            Then we will create fake ICPUBuffers that point to srcBuffer and stagingBuffer with correct offsets
+            Then we have to set the buffer and regions for each one of those ICPUImages using setBufferAndRegions
+            Finally we fill the filter state and `execute` which require in/out CPUImages
+        */
+            
+        auto dstImageParams = dstImage->getCreationParameters();
+
+        // inCPUImage is an image matching the params of dstImage but with the extents and layer count of the current region being copied and mipLevel 1u and the format being srcImageFormat
+        // the buffer of this image is set to (srcBuffer+Offset) and the related region is set to cover the whole copy region (offset from 0)
+        {
+            auto inCpuImageRegionsDynArray = core::make_refctd_dynamic_array<core::smart_refctd_dynamic_array<asset::ICPUImage::SBufferCopy>>(1);
+            auto& inCpuImageRegion = inCpuImageRegionsDynArray->front();
+            inCpuImageRegion = {};
+            inCpuImageRegion.bufferOffset = 0u;
+            inCpuImageRegion.bufferRowLength = mainRegion.bufferRowLength;
+            inCpuImageRegion.bufferImageHeight = mainRegion.bufferImageHeight;
+            inCpuImageRegion.imageSubresource.aspectMask = mainRegion.imageSubresource.aspectMask;
+            inCpuImageRegion.imageSubresource.mipLevel = 0u;
+            inCpuImageRegion.imageSubresource.baseArrayLayer = 0u;
+            inCpuImageRegion.imageOffset.x = 0u;
+            inCpuImageRegion.imageOffset.y = 0u;
+            inCpuImageRegion.imageOffset.z = 0u;
+            inCpuImageRegion.imageExtent.width    = regionToCopyNext.imageExtent.width;
+            inCpuImageRegion.imageExtent.height   = regionToCopyNext.imageExtent.height;
+            inCpuImageRegion.imageExtent.depth    = regionToCopyNext.imageExtent.depth;
+            inCpuImageRegion.imageSubresource.layerCount = core::max(regionToCopyNext.imageSubresource.layerCount, 1u);
+
+            auto localImageOffset = core::vector4du32_SIMD(currentBlockInRow, currentRowInSlice, currentSliceInLayer, currentLayerInRegion);
+            uint64_t offsetInCPUBuffer = mainRegion.bufferOffset + core::dot(localImageOffset, srcBufferByteStrides)[0];
+            uint8_t* inCpuBufferPointer = const_cast<uint8_t*>(reinterpret_cast<const uint8_t*>(srcBuffer->getPointer()) + offsetInCPUBuffer);
+            asset::ICPUImage::SCreationParams inCPUImageParams = dstImageParams;
+            inCPUImageParams.flags = asset::IImage::ECF_NONE; // Because we may want to write to first few layers of CUBEMAP (<6) but it's not valid to create an Cube ICPUImage with less that 6 layers.
+            inCPUImageParams.format = srcImageFormat;
+            inCPUImageParams.extent = regionToCopyNext.imageExtent;
+            inCPUImageParams.arrayLayers = regionToCopyNext.imageSubresource.layerCount;
+            inCPUImageParams.mipLevels = 1u;
```

Comment:
just make a full image with the full extent, mip pyramid and layers and use the `regions` as they are